File: internal/content/d2legacy/lua/d2legacy/gameplay/targeting_test.lua
local test = require("d2legacy.tests/v1")

return test.suite({
    profile = "ecs",
    tier = "fast",
    cases = {
        test.case("selects_by_priority_distance_and_stable_id", {
            test.run(function()
                local ecs = require("engine.ecs/v1")
                local function spawn(id, kind, x, priority)
                    ecs.create({
                        ["d2legacy.world.position"] = { x = x, y = 10 },
                        ["d2legacy.world.selectable"] = {
                            id = id,
                            kind = kind,
                            label = id,
                            owner = "owner:" .. id,
                            radius = 3,
                            priority = priority,
                        },
                    })
                end
                spawn("near", "npc", 10, 0)
                spawn("far", "npc", 12, 0)
                spawn("item", "item", 12, 5)
                spawn("distance-near", "npc", 20, 0)
                spawn("distance-far", "npc", 22, 0)
                spawn("stable-b", "npc", 30, 0)
                spawn("stable-a", "npc", 30, 0)
                local targeting = require("d2legacy.gameplay.targeting")
                local hit = targeting.selectable_at(10, 10)
                test.assert(
                    hit and hit.id == "item" and hit.kind == "item" and hit.label == "item",
                    [=[hit and hit.id == "item" and hit.kind == "item" and hit.label == "item"]=]
                )
                test.assert(
                    hit.owner == "owner:item" and hit.x == 12 and hit.y == 10 and hit.radius == 3,
                    [=[hit.owner == "owner:item" and hit.x == 12 and hit.y == 10 and hit.radius == 3]=]
                )
                test.assert(
                    targeting.selectable_at(20, 10).id == "distance-near",
                    [=[targeting.selectable_at(20, 10).id == "distance-near"]=]
                )
                test.assert(
                    targeting.selectable_at(30, 10).id == "stable-a",
                    [=[targeting.selectable_at(30, 10).id == "stable-a"]=]
                )
                test.assert(targeting.selectable_at(100, 100) == nil, [=[targeting.selectable_at(100, 100) == nil]=])
                test.assert(targeting.selectable_at(0 / 0, 10) == nil, [=[targeting.selectable_at(0 / 0, 10) == nil]=])
            end),
        }),
    },
})

File: internal/mod/d2legacy/adapter/targeting/targeting_test.go
package targeting

import (
	"github.com/gravestench/akara"
	gameecs "github.com/gravestench/dark-magic/internal/game/ecs"
	"testing"
)

func TestResolverUsesSpawnedEntityKindsAndPriority(t *testing.T) {
	engine := gameecs.New()
	positions, err := akara.RegisterSchema(engine.World(), akara.Schema{Name: "d2legacy.world.position", Fields: []akara.Field{{Name: "x", Kind: akara.FieldFloat64}, {Name: "y", Kind: akara.FieldFloat64}}})
	if err != nil {
		t.Fatal(err)
	}
	selectables, err := akara.RegisterSchema(engine.World(), Schema())
	if err != nil {
		t.Fatal(err)
	}
	spawn := func(id, kind string, priority int64) {
		entity := engine.World().MustCreateEntity()
		if _, err := positions.Set(entity, map[string]any{"x": 10.0, "y": 10.0}); err != nil {
			t.Fatal(err)
		}
		if _, err := selectables.Set(entity, map[string]any{"id": id, "kind": kind, "label": id, "owner": "", "radius": 2.0, "priority": priority}); err != nil {
			t.Fatal(err)
		}
	}
	spawn("npc:1", "npc", 0)
	spawn("item:1", "item", 5)
	hit, found := New(engine).HitAt(10, 10)
	if !found || hit.ID != "item:1" || hit.Kind != "item" {
		t.Fatalf("hit=%#v,%v", hit, found)
	}
}
